src/collection.js

```diff
@@ -121,4 +122,84 @@ Collection.pluck = function pluck (collection$, sinkProperty) {
     .startWith([]);
 };
 
+// convert a stream of items' sources snapshots into a stream of collections
+Collection.gather = function gather (itemsState$ , component, sources, handlers = {}, idAttribute = 'id') {
+  const makeDestroyable = component => (sources) => {
+    const sinks = component(sources);
+    return {
+      ...sinks,
+      remove$: xs.merge(sinks.remove$ || xs.never(), sources.destroy$)
+    };
+  };
+  const collection = Collection(makeDestroyable(component), sources, {
+    remove$: (collection, item) => collection.remove(item),
+    ...handlers
+  });
+  // each time a new item appears, it should be added to the collection
+  const addReducers$ = itemsState$
+    // get the added items at each step
+    .fold(
+      ({prevIds}, items) => ({
+        prevIds: items.map(item => item[idAttribute]),
+        addedItems: items.filter(item => prevIds.indexOf(item[idAttribute]) === -1)
+      }),
+      {
+        prevIds: [],
+        addedItems: []
+      }
+    )
+    .map(({addedItems}) => addedItems)
+    .filter(addedItems => addedItems.length)
+    // turn each new item into a hash of source streams, tracking all the future updates
+    .map(addedItems =>
+      addedItems.map(addedItem => {
+        const itemStateInfinite$ = itemsState$
+          .map(items =>
+            items.find(item => item[idAttribute] === addedItem[idAttribute])
+          );
+        // if an item isn't present if a new snapshot, it shall be destroyed
+        const destroy$ = itemStateInfinite$.filter(item => !item).take(1);
+        const itemState$ = itemStateInfinite$.endWhen(destroy$);
+        
+        return Object.keys(addedItem)
+          .reduce((sources, key) => {
+            if (key === idAttribute) {
+              return sources;
+            }
+
+            return {
+              ...sources,
+              [key]: itemState$
+                .map(state => state[key])
+                .startWith(addedItem[key])
+                // skip the snapshot if the value didn't change
+                .compose(dropRepeats((value, nextValue) => {
```

Comment:
Can we also pull this inner arrow function out into a named function?

It's quite deeply nested so it's hard to see exactly what's happening.